File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
from collections import Counter
def solution(phone_book):
    _li = [set() for _ in range(21)]
    phone_book.sort(key = lambda x: len(x))
    min_len = len(phone_book[0])
    _li[len(phone_book[0])].add(phone_book[0])
    
    for i in range(1, len(phone_book)):
        for j in range(len(phone_book[i]) - min_len):
            if phone_book[i][:min_len + j] in _li[min_len + j]:
                return False
        _li[len(phone_book[i])].add(phone_book[i])
    return True

File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        Set<String> set = new HashSet<>();
        Arrays.sort(phone_book, new Comparator<String>(){
            public int compare(String s1, String s2){
                return s1.length() - s2.length();
            }
        });
        for(String phone_num: phone_book){
            if(set.size() == 0)
                set.add(phone_num);
            else{
                for(int i = 1; i< phone_num.length() + 1; ++i){
                    if(set.contains(phone_num.substring(0, i)))
                        return false;
                }
                set.add(phone_num);
            }
        }
        return true;
    }
}